### PR TITLE
Adds `fused_all_reduce()` to `physicsnemo.distributed`

### PR DESCRIPTION
**Pull request by peterdsharpe**
# PhysicsNeMo Pull Request

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/distributed/utils.py [hunk 1]
```diff
@@ -150,6 +153,260 @@ def reduce_loss(loss: float, dst_rank: int = 0, mean: bool = True):  # pragma: n
...
+
+    # Pack: flatten + cast every tensor and concatenate into ONE buffer.  ``cat``
+    # of flattened tensors (not ``stack``) tolerates heterogeneous leaf shapes,
+    # and always allocates, so the buffer never aliases the inputs.
+    flats = [t.reshape(-1).to(device=buffer_device, dtype=work_dtype) for t in detached]
+    numels = [f.numel() for f in flats]
+    buffer = torch.cat(flats)
+
+    # The one collective.
+    dist.all_reduce(buffer, op=op, group=group)
```
Comment (coreyjadams) [inline physicsnemo/distributed/utils.py]: If we

Comment (coreyjadams) [inline physicsnemo/distributed/utils.py]: Whoops, previous comment was a flub.

I was trying to say:

If we want to be able to compile this, we may want to use the functional collectives.  I don't think it's necessary though.

File: physicsnemo/distributed/fused_all_reduce.py [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+        """
+        if self._completed:
+            return True
+        self._work.wait()
+        # Deferred unpack: split the reduced buffer and cast each slice back to
+        # its leaf's dtype/device (``copy_`` handles both).
+        with torch.no_grad():
+            offset = 0
+            for out, n in zip(self._outputs, self._numels):
+                out.copy_(self._buffer[offset : offset + n].reshape(out.shape))
```
Comment (coreyjadams) [inline physicsnemo/distributed/fused_all_reduce.py:96]: Who makes the check that buffer is large enough, before this?

File: physicsnemo/distributed/fused_all_reduce.py [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+    Returns
+    -------
+    tuple[list[torch.Tensor], _FusedAllReduceWork | None]
+        The reduced tensors (same order as the input) and a work handle. The
+        handle is ``None`` when ``async_op`` is ``False``; otherwise it is a
+        :class:`_FusedAllReduceWork` whose ``wait`` completes the collective and
+        fills the outputs (which hold the local, un-reduced values until then).
+    """
+    if not tensors:
+        return [], (_FusedAllReduceWork() if async_op else None)
```
Comment (coreyjadams) [inline physicsnemo/distributed/fused_all_reduce.py:201]: I like there is a path to get an Async handle to a list of empty futures.  There is something very 2026 to asynchronously getting an empty future back from your code.

File: physicsnemo/distributed/fused_all_reduce.py [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+        return [t.clone() for t in detached], (
+            _FusedAllReduceWork() if async_op else None
+        )
+
+    buffer_device = torch.device(device) if device is not None else detached[0].device
+
+    # Pack: flatten + cast every tensor and concatenate into ONE buffer.  ``cat``
+    # of flattened tensors (not ``stack``) tolerates heterogeneous leaf shapes,
+    # and always allocates, so the buffer never aliases the inputs.
+    flats = [t.reshape(-1).to(device=buffer_device, dtype=work_dtype) for t in detached]
```
Comment (coreyjadams) [inline physicsnemo/distributed/fused_all_reduce.py:244]: Ideally, you're doing no-ops to transfer devices here.  If you're moving across devices then this whole fusion is pointless IMO.

But anyways, *IF* you're going to have possible transfers, reorder this?  Do the concatenate (since we're fusing small tensors) and then transfer once.

File: physicsnemo/distributed/fused_all_reduce.py [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+        # Launch non-blocking and defer the unpack to the handle's wait().
+        # Outputs start as clones of the local (un-reduced) leaves, so a
+        # premature read is stale-but-valid (matching dist.all_reduce's in-place
+        # semantics), never uninitialized garbage; wait() overwrites them.
+        outputs = [s.clone() for s in detached]
+        work = dist.all_reduce(buffer, op=op, group=group, async_op=True)
+        return outputs, _FusedAllReduceWork(work, buffer, outputs, numels)
+
+    # The one (blocking) collective.
+    dist.all_reduce(buffer, op=op, group=group)
```
Comment (coreyjadams) [inline physicsnemo/distributed/fused_all_reduce.py]: Does this not launch the all_reduce twice?

Comment (coreyjadams) [reply physicsnemo/distributed/fused_all_reduce.py]: Wait, I missed the return and was confused by the "one (blocking) collective".

**Review by coreyjadams — APPROVED** [review]
Looking good.  I think it might be nice to add some comments in the docstring about intended use cases like "this is good in a training loop to globally reduce your metrics without stalling your code and killing your scaling efficiency, etc" but up to you!

File: physicsnemo/distributed/fused_all_reduce.py (hunk 5 above)
Comment (peterdsharpe) [reply physicsnemo/distributed/fused_all_reduce.py]: Rewrote this to have a cleaner `if: else:` :)

File: physicsnemo/distributed/fused_all_reduce.py [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+    # structure instead of silently degrading it to a plain dict.
+    if isinstance(tensors, TensorDictBase):
+        leaves = list(tensors.items(include_nested=True, leaves_only=True))
+        keys = [key for key, _ in leaves]
+        reduced, work = _reduce_keyed(
+            keys, [value for _, value in leaves], reduce_leaves
+        )
+        # Clone the input as a structure/dtype/device template, then write each
+        # reduced leaf back by its (possibly nested) key.
+        out = tensors.detach().clone()
```
[the remark this answers is not in the document]
Comment (peterdsharpe) [reply physicsnemo/distributed/fused_all_reduce.py]: Not an actual issue; TensorDict also preserves identity on assign

File: physicsnemo/distributed/fused_all_reduce.py (hunk 2 above)
Comment (peterdsharpe) [reply physicsnemo/distributed/fused_all_reduce.py:96]: So, in the one place this is used, it is correct by construction (comes directly from `flats`, which is synced with `numels` that also come from `flats`).

That said I could add an `if` check here (perhaps behind a is-not-compiling guard) inside the `_FusedAllReduceWork()`, that's pretty cheap. Will do

File: physicsnemo/distributed/utils.py (hunk 1 above)
Comment (peterdsharpe) [reply physicsnemo/distributed/utils.py]: I didn't even know there were functional collectives now, cool! Will give this is a shot, an easy one-line swap

Comment (peterdsharpe) [reply physicsnemo/distributed/utils.py]: Actually, upon testing: on `torch 2.12`, `dist.all_reduce` does not graph-break under `fullgraph=True`. Maybe this is fine then?

Still, I have a slight preference for functional interfaces over side-effect ones, so I'll try the functional method

File: physicsnemo/distributed/fused_all_reduce.py (hunk 4 above)
Comment (peterdsharpe) [reply physicsnemo/distributed/fused_all_reduce.py:244]: Good call

File: test/distributed/test_fused_all_reduce.py [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+    )
+
+    # Deferred unpack: until wait(), the outputs are this rank's local clones.
+    assert result["scalar"].item() == pytest.approx(float(rank + 1))
+
+    assert work.wait() is True
+    # After wait(), the fused SUM is filled into the outputs in place.
+    assert result["scalar"].item() == pytest.approx(expected)
+    assert torch.allclose(result["vec"], torch.full((3,), expected, device=dm.device))
+    assert work.is_completed()
```
[the remark this answers is not in the document]
Comment (peterdsharpe) [reply test/distributed/test_fused_all_reduce.py:541]: added

File: physicsnemo/distributed/fused_all_reduce.py [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+        return (out, work) if async_op else out
+
+    if isinstance(tensors, Mapping):
+        keys = list(tensors.keys())
+        reduced, work = _reduce_keyed(
+            keys, [tensors[key] for key in keys], reduce_leaves
+        )
+        result = dict(zip(keys, reduced))
+        return (result, work) if async_op else result
+
```
[the remark this answers is not in the document]
Comment (peterdsharpe) [reply physicsnemo/distributed/fused_all_reduce.py]: added

**Comment by peterdsharpe** [comment]
/ok to test https://github.com/NVIDIA/physicsnemo/pull/1761/commits/265f2cc5ef63dd3ac7f985b93b3862d19218b253
